test/format/replay.c

```diff
@@ -0,0 +1,467 @@
+ * the same timestamp will always have their data compare identically. Predictable replay only works
+ * with timestamped transactions and to avoid complexity, only a single operation is allowed in a
+ * transaction.
+ *
+ * To achieve the predictability we use two random number generators (the data RNGs and the extra
+ * RNG) with known start seeds, the data seed and the extra seed. Every single-threaded modification
+ * (like bulk loading) when deciding on a random course, uses the global data RNG, which is seeded
+ * by the data seed. Global decisions that don't affect data, like whether to turn on verbose, or
+ * even the rate of checkpointing, use the global "extra" RNG, which is seeded by the "extra seed".
+ * Changing the extra seed may change some characteristics of how a workload is tested, but should
+ * not change any data on disk. When worker threads run, they have their own data and extra RNGs,
+ * and these are seeded by the timestamp they are working on.
+ *
+ * Before a worker thread can decide on what operation to do on which key in which table, it must
+ * obtain the next timestamp. Timestamps are doled out atomically, so no two worker thread can ever
+ * "work on" the same timestamp. The timestamp is XOR-ed with the data seed, the result is the seed
+ * of the thread's private data RNG for the duration of that operation. Likewise, a private extra
+ * RNG is seeded from the timestamp and the extra seed. This ensures that all decisions about what
+ * is committed at that timestamp is predictable based on the timestamp. As you might expect, the
+ * thread's data RNG is used to decide what operation to do, which table to use, and which key
+ * within the table. Other "random" decisions, like whether to reopen a session, or whether to
+ * repeat a read from the snap list, use the extra RNG.
+ *
+ * Note that once a thread has started to work on an operation at a timestamp, it cannot give up on
+ * the effort. If, for example, a rollback error naturally happens, we can rollback the transaction.
+ * However, immediately getting a new timestamp would mean that we would lose the consequences of
+ * the previous timestamp, perhaps a record would not be updated in a particular way. Thus, after a
+ * rollback, a thread starts again, using the same timestamp it had before, and it seeds its RNGs
+ * again using this timestamp. This gives full predictability, even in the face of temporary
+ * failures.
+ *
+ * To avoid the possibility that two threads work on the same key at the same time, we have the
+ * concept of "lanes", and only one thread can be working in a lane at once. There are LANE_COUNT
+ * lanes, where LANE_COUNT is 2^k for some k. A thread uses a data RNG to choose the top bits of a
+ * key number, but the bottom k bits of the key number are set to the bottom k bits of the timestamp
+ * being worked. Those bottom k bits also determine the "lane" we are in. Each lane has a flag that
```

Comment:
"being worked" is awkward. Maybe '...of the timestamp being used in this transaction.'?
And again, un-quote "lane" here.